File: backend/app/services/collection_service.py
from fastapi import Depends, HTTPException

from sqlalchemy import Table, inspect, Column
from sqlalchemy.orm import Session
from ..database.database import metadata, get_db, SQLALCHEMY_TYPE_MAP, get_table, Session
from ..schemas.schemas import CollectionCreateRequest

class CollectionService:
    def __init__(self, db: Session = Depends(get_db)):
        self.db = db

    def list(self):
        try:
            inspector = inspect(self.db.bind)
            tables = inspector.get_table_names()
            return {"collections": tables}
        except Exception as e:
            raise HTTPException(status_code=500, detail=f"Failed to list collections: {e}")
        
    def create(self, request: CollectionCreateRequest):
        try:
            # Check if table already exists
            inspector = inspect(self.db.bind)
            if request.table_name in inspector.get_table_names():
                raise HTTPException(
                    status_code=400, 
                    detail=f"Collection '{self.request.table_name}' already exists"
                )
            
            # Build columns programmatically
            columns = []

            for col_def in self.request.columns:
                if col_def.type not in SQLALCHEMY_TYPE_MAP:
                    raise HTTPException(
                        status_code=400,
                        detail=f"Unsupported column type: {col_def.type}"
                    )
                
                sqlalchemy_type = SQLALCHEMY_TYPE_MAP[col_def.type]
                
                column = Column(
                    col_def.name,
                    sqlalchemy_type,
                    nullable=col_def.nullable,
                    primary_key=col_def.primary_key
                )
                columns.append(column)
            
                
            # Create table using SQLAlchemy
            new_table = Table(request.table_name, metadata, *columns)
            new_table.create(self.db.bind)
            
            return {
                "message": f"Collection '{request.table_name}' created successfully",
                "columns": [{"name": col.name, "type": str(col.type)} for col in columns]
            }
        except HTTPException:
            raise
        except Exception as e:
            raise HTTPException(status_code=500, detail=f"Failed to create collection: {e}")
        
    def columns(self, collection_name: str):
        try:
            table = get_table(collection_name, self.db)

            columns = [
                {
                    "name": col.name,
                    "type": str(col.type),
                    "nullable": col.nullable,
                    "primary_key": col.primary_key
                }
                for col in table.columns
            ]

            return {"columns": columns}
        except HTTPException:
            raise
        except Exception as e:
            raise HTTPException(status_code=500, detail=f"Failed to get columns: {e}")
        
    def delete(self, collection_name: str):
        try:
            table = get_table(collection_name, self.db)
            table.drop(self.db.bind)
            
            # Remove from metadata to avoid conflicts
            if collection_name in metadata.tables:
                metadata.remove(metadata.tables[collection_name])
            
            return {"message": f"Collection '{collection_name}' deleted successfully"}
        except HTTPException:
            raise
        except Exception as e:
            raise HTTPException(status_code=500, detail=f"Failed to delete collection: {e}")

File: frontend/app/src/lib/sdk/services/CollectionService.ts
import Bird, { SendMethod } from '../Bird';

type ColumnDefinition = {
  name: string;
  type: string;
  nullable?: boolean;
  primary_key: boolean;
};

export default class CollectionService {
  bird: Bird;
  baseUrl: string;

  constructor(bird: Bird) {
    this.bird = bird;
    this.baseUrl = `${this.bird.url}/api/collections`;
  }

  async list() {
    try {
      const data = await this.bird.send(this.baseUrl, {
        method: SendMethod.GET,
      });

      return data.collections;
    } catch (e) {
      console.log(e);
    }
  }

  async create(table_name: string, columns: ColumnDefinition[], type: string) {
    try {
      const data = await this.bird.send(this.baseUrl, {
        method: SendMethod.POST,
        body: {
          table_name: table_name,
          fields: columns,
          type: type,
        },
      });

      return data.collection;
    } catch (e) {
      console.log(e);
    }
  }

  async columns(collectionName: string) {
    try {
      const data = await this.bird.send(`${this.baseUrl}/${collectionName}`, {
        method: SendMethod.GET,
      });

      return data.columns;
    } catch (e) {
      console.log(e);
    }
  }

  async delete(collectionName: string) {
    try {
      const data = await this.bird.send(`${this.baseUrl}/${collectionName}`, {
        method: SendMethod.DELETE,
      });

      return data.message;
    } catch (e) {
      console.log(e);
    }
  }
}
